### ISSUE TITLE
[FSDP] Generalize `summon_full_params()` to non-FSDP root module

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#74165]
We should generalize the method `summon_full_params()` to a static method to accommodate being called on a root module that is not an FSDP instance. This avoids having to search for FSDP root instances within the root module and aggregating manually.

For example, right now if we want to compare the parameters of `root_module` with a fully local version of the model `local_model`, then we may have to do something like:
```
params = []
for module in root_module.children():
    context = module.summon_full_params() if isinstance(module, FSDP) else suppress()
    with context:
        params.extend(list(p.detach().clone() for p in module.parameters())
# Now we can correctly call `zip(params, local_model.parameters())`
```
We should place this kind of logic behind our API. The generalization should be similar to the above only using `yield` instead of explicitly constructing a `list` and not `detach().clone()`-ing since the memory should be safe to access while inside the context.

cc @zhaojuanmao @mrshenli @rohan-varma

**Comment by rohan-varma** [comment]
fixed now
